### PR TITLE
Overlappingsubnet

### PR DESCRIPTION
**Pull request by gandhihardikm**
Fix for overlapping networks in a tenant allowed (#440)

### CODE REVIEW

File: netmaster/master/network.go [hunk 1]
```diff
@@ -95,6 +95,7 @@ func CreateNetwork(network intent.ConfigNetwork, stateDriver core.StateDriver, t
 	}
 
 	subnetIP, subnetLen, _ := netutils.ParseCIDR(network.SubnetCIDR)
+
```
Comment (gaurav-dalvi) [inline netmaster/master/network.go]: Do we really need this in this PR ? 
@shaleman Could you comment on whitespaces and new lines addition in a change. I guess gofmt should take care of this.

File: netmaster/objApi/objapi_test.go [hunk 2]
```diff
@@ -712,6 +711,101 @@ func TestTenantAddDelete(t *testing.T) {
...
+	checkCreateTenant(t, false, "tenant1")
+	checkCreateNetwork(t, false, "tenant1", "contiv1", "", "vlan", "10.1.0.0/16", "", 2, "ffff:ffff:ffff::/48", "")
+	checkInspectNetwork(t, false, "default", "contiv", "", 1, 0)
+	checkInspectNetwork(t, false, "tenant1", "contiv1", "", 2, 0)
+	checkDeleteNetwork(t, false, "default", "contiv")
+	checkDeleteNetwork(t, false, "tenant1", "contiv1")
+	checkDeleteTenant(t, false, "tenant1")
+
+}
+
```
Comment (gaurav-dalvi) [inline netmaster/objApi/objapi_test.go]: Good Test cases !

File: netmaster/master/network.go (hunk 1 above)
Comment (shaleman) [inline netmaster/master/network.go]: yes, please remove white space changes

File: netmaster/objApi/apiController.go [hunk 3]
```diff
@@ -663,6 +664,32 @@ func (ac *APIController) NetworkCreate(network *contivModel.Network) error {
 		return core.Errorf("Tenant not found")
 	}
 
+	for key := range tenant.LinkSets.Networks {
+		networkDetail := contivModel.FindNetwork(key)
+		if networkDetail == nil {
+			log.Errorf("Network %s not found", network)
+			return errors.New("FromNetwork not found")
```
Comment (shaleman) [inline netmaster/objApi/apiController.go]: please fix the error message

File: netmaster/objApi/apiController.go [hunk 4]
```diff
@@ -663,6 +664,32 @@ func (ac *APIController) NetworkCreate(network *contivModel.Network) error {
...
+		}
+
+		// Check for overlapping subnetv6 if existing and current subnetv6 is non-empty
+		if network.Ipv6Subnet != "" && networkDetail.Ipv6Subnet != "" {
+			flagv6 := netutils.IsOverlappingSubnetv6(network.Ipv6Subnet, networkDetail.Ipv6Subnet)
+			if flagv6 == true {
+				log.Errorf("Overlapping of Subnetv6 Networks")
+				return core.Errorf("Network " + networkDetail.NetworkName + " conflicts with subnetv6  " + network.Ipv6Subnet)
```
Comment (shaleman) [inline netmaster/objApi/apiController.go]: please use errors.New()
formatting for core.Errorf() is not user friendly

File: utils/netutils/netutils.go [hunk 5]
```diff
@@ -48,8 +50,105 @@ func init() {
 	}
 }
 
+func ipv6ToInt(IPv6Addr string) *big.Int {
```
Comment (shaleman) [inline utils/netutils/netutils.go]: Please call this func ipv6ToBigint()

File: utils/netutils/netutils.go [hunk 6]
```diff
@@ -48,8 +50,105 @@ func init() {
...
+	inputIPv6StartRange, inputIPv6EndRange := getIPv6Range(inputIPv6Subnet)
+	existingIPv6StartRange, existingIPv6EndRange := getIPv6Range(existingIPv6Subnet)
+
+	inputStartRange := ipv6ToInt(inputIPv6StartRange)
+	inputEndRange := ipv6ToInt(inputIPv6EndRange)
+
+	existingStartRange := ipv6ToInt(existingIPv6StartRange)
+	existingEndRange := ipv6ToInt(existingIPv6EndRange)
+
+	if (existingStartRange.Cmp(inputStartRange) == -1 && inputStartRange.Cmp(existingEndRange) == -1) ||
```
Comment (shaleman) [inline utils/netutils/netutils.go]: I think this logic can be simplified using `>= 0` or `<= 0`

File: utils/netutils/netutils.go [hunk 7]
```diff
@@ -48,8 +50,105 @@ func init() {
...
+
+	if (inputStartRange.Cmp(existingStartRange) == -1 && existingStartRange.Cmp(inputEndRange) == -1) ||
+		(inputStartRange.Cmp(existingEndRange) == -1 && existingEndRange.Cmp(inputEndRange) == -1) {
+		return true
+	}
+	return false
+}
+
+// String returns the hexadecimal form of m, with no punctuation.
+func (m ipAddress) byteToString() string {
```
Comment (shaleman) [inline utils/netutils/netutils.go]: Please see if you can reuse an existing function from golfing `net` pkg.
I'm not convinced you need a new type for `ipAddress`. you should be able to reuse `net.IP` type

File: netmaster/objApi/objapi_test.go (hunk 2 above)
Comment (shaleman) [inline netmaster/objApi/objapi_test.go]: 👍

**Comment by shaleman** [comment]
Largely looks good other than minor comments.

**Comment by shaleman** [comment]
LGTM

**Pull request merged**